src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/ILspServices.cs

```diff
@@ -0,0 +1,19 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Immutable;
+
+namespace CommonLanguageServerProtocol.Framework;
+
+// TODO: Does this need a GetRequiredServices<T> or is that too specific to some implementations?
+public interface ILspServices : IDisposable
+{
+    T GetRequiredService<T>();
+
+    bool TryGetService(Type @type, out object service);
+
+    ImmutableArray<Type> GetRegisteredServices();
```

Comment:
what's the value of this method?

Reply:
Thanks for pointing this one out, I'd be happy to get rid of it if possible.

[Here's where it gets used now](https://github.com/dotnet/roslyn/pull/62924/files#diff-662475d536e197718f627878b41d88abb21219ae3dc7dac698ffeb6a81267545R157). The trick is that Roslyn's LspServices implementation does not seem to support registering implementors as the same type. IE, [`HoverHandler`](https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/Core/LanguageServer/Handlers/Hover/HoverHandler.cs) is not registered as some `IRequestHandler` instance, it's registered as `HoverHandler` and then the individual services get examined to see if they match the type we want. That's obviously not ideal but it reflects how Roslyn does this today. The one benefit I see to it is that it allows the Handler construction to be [actually Lazy](https://github.com/dotnet/roslyn/pull/62924/files#diff-662475d536e197718f627878b41d88abb21219ae3dc7dac698ffeb6a81267545R168) where-as the `GetRequiredServices` model does not.

CC @dibarbet who has been my guide on Roslyn's Services model so far. Ideally I'd like to get rid of `GetRegisteredServices`, `SupportsGetRegisteredServices`, and `SupportsGetRequiredServices` on this interface and require each implementation's `GetRequiredServices` to behave as it does for MS.Ext.DependencyInjection (even if that means internally you have to do the type reflection).

Reply:
Maybe leave this as a roslyn impl detail?  Unfortunately its pretty important for the handlers and LSP services to be lazy to avoid loading a bunch of things until they are used.  

I did try exporting as IRequestHandler + the type information metadata, but then MEF loads those assemblies while scanning them.  And unfortunately we can't reflect on the actual type unless you actually create it or export it as that type.